synapse/storage/databases/main/event_federation.py

```diff
@@ -1583,6 +1583,49 @@ def _record_event_failed_pull_attempt_upsert_txn(
 
         txn.execute(sql, (room_id, event_id, 1, self._clock.time_msec(), cause))
 
+    @trace
+    async def separate_event_ids_with_failed_pull_attempts(
+        self, event_ids: Collection[str]
+    ) -> Tuple[Collection[str], Collection[str]]:
```

Comment:
Use `StrCollection`, see https://github.com/matrix-org/synapse/issues/14809

Reply:
Tracking on non-outdated portion of the diff in https://github.com/matrix-org/synapse/pull/15585#discussion_r1197029802